test/go: TestRuntimeSetFinalizerTinyObjects is flaky

## Summary

`TestRuntimeSetFinalizerTinyObjects` can fail intermittently in LLGo CI when too few finalizers run before its deadline.

## Observed failure

The failure occurred in PR #2272, whose changes are limited to the GOROOT test runner and its tests.

Failed job: https://github.com/xgo-dev/llgo/actions/runs/30794803798/job/91625841291

Environment:

- Ubuntu latest
- LLVM 19
- Go 1.24.2
- LLGo test shard 0

```text
--- FAIL: TestRuntimeSetFinalizerTinyObjects (3.00s)
    finalizer_test.go:61: only 12/32 finalizers ran
FAIL github.com/goplus/llgo/test/go
```

The test creates 32 `*int32` objects with finalizers and requires more than half of the finalizers to run within three seconds while repeatedly calling `runtime.GC()`.

Other LLGo jobs in the same workflow completed successfully, including the Go 1.24.2 Ubuntu shard 1 job and both Go 1.26.5 Ubuntu shards.

## Reproduction status

A focused local stress run at the PR head on Darwin/arm64 with Go 1.24.0 completed 50 iterations successfully:

```text
llgo test -run "^TestRuntimeSetFinalizerTinyObjects$" -count=50 -timeout=10m ./test/go
PASS
```

There is currently no deterministic reproducer; the failure appears timing- or environment-dependent.

--- a/runtime/internal/clite/bdwgc/bdwgc.go
+++ b/runtime/internal/clite/bdwgc/bdwgc.go
@@ -88,6 +88,12 @@ func RegisterFinalizerUnreachable(
 	fn FinalizerFunc, cd c.Pointer,
 	oldFn *FinalizerFunc, oldCd *c.Pointer)
 
+// InvokeFinalizers runs all finalizers that BDWGC has queued as ready.
+// It returns the number of finalizers that were run.
+//
+//go:linkname InvokeFinalizers C.GC_invoke_finalizers
+func InvokeFinalizers() c.Int
+
 // -----------------------------------------------------------------------------
 
 //go:linkname Enable C.GC_enable

--- a/runtime/internal/lib/runtime/runtime_gc.go
+++ b/runtime/internal/lib/runtime/runtime_gc.go
@@ -36,17 +36,23 @@ func ReadMemStats(m *runtime.MemStats) {
 }
 
 func GC() {
-	bdwgc.Gcollect()
-	runFinalizers()
-	// BDW finalizers are observed on a subsequent collection cycle.
+	collectAndRunFinalizers()
 	// Run one extra cycle so weak-pointer cleanup hooks (unique/weak) see
 	// finalized state before we trigger map cleanup callbacks.
-	bdwgc.Gcollect()
-	runFinalizers()
+	collectAndRunFinalizers()
 	unique_runtime_notifyMapCleanup()
 	if poolCleanup != nil {
 		poolCleanup()
 	}
+}
+
+func collectAndRunFinalizers() {
+	bdwgc.Gcollect()
+	// GC_gcollect only discovers unreachable finalizable objects. Explicitly
+	// drain BDWGC's ready queue so runtime.GC does not depend on a later
+	// allocation to invoke the callbacks that feed runFinalizers.
+	bdwgc.InvokeFinalizers()
+	runFinalizers()
 }
 
 func saturatingSub(x, y uintptr) uintptr {

--- a/test/go/finalizer_llgo_test.go
+++ b/test/go/finalizer_llgo_test.go
@@ -1,0 +1,50 @@
+//go:build llgo && !baremetal && !nogc
+
+package gotest
+
+import (
+	"runtime"
+	"testing"
+	_ "unsafe"
+)
+
+//go:linkname getBDWGCFinalizeOnDemand C.GC_get_finalize_on_demand
+func getBDWGCFinalizeOnDemand() int32
+
+//go:linkname setBDWGCFinalizeOnDemand C.GC_set_finalize_on_demand
+func setBDWGCFinalizeOnDemand(enabled int32)
+
+func TestRuntimeGCDrainsBDWGCFinalizersOnDemand(t *testing.T) {
+	// BDWGC normally may invoke ready finalizers during a later allocation.
+	// On-demand mode makes runtime.GC's explicit drain observable without
+	// relying on allocation timing. This setting is process-global, so this
+	// test and the neighboring finalizer tests must remain sequential.
+	old := getBDWGCFinalizeOnDemand()
+	setBDWGCFinalizeOnDemand(1)
+	t.Cleanup(func() {
+		setBDWGCFinalizeOnDemand(old)
+	})
+
+	const n = 32
+	finalized := make(chan int32, n)
+	created := make(chan struct{})
+	go func() {
+		makeFinalizerTinyObjects(n, finalized)
+		close(created)
+	}()
+	<-created
+
+	// InvokeFinalizers runs synchronously once BDWGC has queued an object.
+	// The retries only let the producer goroutine exit so that conservative
+	// stack and register roots no longer keep the objects reachable.
+	for range 8 {
+		runtime.Gosched()
+		runGCWithTimeout(t)
+		if len(finalized) > n/2 {
+			return
+		}
+	}
+	if got := len(finalized); got <= n/2 {
+		t.Fatalf("runtime.GC ran only %d/%d on-demand finalizers", got, n)
+	}
+}

